server/src/main/java/org/elasticsearch/ingest/IngestDocument.java

```diff
@@ -707,26 +719,31 @@ private Map<String, Object> createTemplateModel() {
     }
 
     /**
-     * one time operation that extracts the metadata fields from the ingest document and returns them.
-     * Metadata fields that used to be accessible as ordinary top level fields will be removed as part of this call.
+     * Get source and metadata map
      */
-    public Map<Metadata, Object> extractMetadata() {
-        Map<Metadata, Object> metadataMap = new EnumMap<>(Metadata.class);
-        for (Metadata metadata : Metadata.values()) {
-            metadataMap.put(metadata, sourceAndMetadata.remove(metadata.getFieldName()));
-        }
-        return metadataMap;
+    public Map<String, Object> getSourceAndMetadata() {
+        return sourceAndMetadata;
     }
 
     /**
-     * Does the same thing as {@link #extractMetadata} but does not mutate the map.
+     * Get source and metadata map as {@link IngestSourceAndMetadata}
      */
-    public Map<Metadata, Object> getMetadata() {
-        Map<Metadata, Object> metadataMap = new EnumMap<>(Metadata.class);
-        for (Metadata metadata : Metadata.values()) {
-            metadataMap.put(metadata, sourceAndMetadata.get(metadata.getFieldName()));
-        }
-        return metadataMap;
+    public IngestSourceAndMetadata getIngestSourceAndMetadata() {
+        return sourceAndMetadata;
+    }
+
+    /**
+     * Get all Metadata values in a Map
+     */
+    public Map<String, Object> getMetadata() {
```

Comment:
Can users get call getSourceAndMetadata() and call the appropriate method instead of adding access here?

Reply:
`org.elasticsearch.action.ingest.WriteableIngestDocument` can't see the package private `org.elasticsearch.ingest.IngestSourceAndMetadata`